test-data/unit/check-statements.test

```diff
@@ -2231,6 +2231,32 @@ class B: pass
 def foo(x: int) -> Union[Generator[A, None, None], Generator[B, None, None]]:
     yield x  # E: Incompatible types in "yield" (actual type "int", expected type "Union[A, B]")
 
+[case testYieldFromUnionOfGenerators]
+from typing import Generator, Union
+
+class T: pass
+
+def foo(arg: Union[Generator[int, None, T], Generator[str, None, T]]) -> Generator[Union[int, str], None, T]:
+    return (yield from arg)
+
+[case testYieldFromInvalidUnionReturn]
+from typing import Generator, Union
+
+class T: pass
+class R: pass
+
+def foo(arg: Union[T, R]) -> Generator[Union[int, str], None, T]:
+    return (yield from arg) # E: "yield from" can't be applied to "Union[T, R]"
+
+[case testYieldFromUnionOfGeneratorWithIterableStr]
+from typing import Generator, Union, Iterable, Optional
+
+class T: pass
+
+def foo(arg: Union[Generator[int, None, T], Iterable[str]]) -> Generator[Union[int, str], None, Optional[T]]:
```

Comment:
This is unsound and should arguably error. For example, I could pass a `Generator[str, None, str]` and get a different type out.

Possibly we should allow it for pragmatic reasons, but then again I'm not sure how often people use `yield from` with values typed as just `Iterable`.

Reply:
Oh this is a great point. I've added a test, but I think this is basically pre-existing. mypy's existing behaviour is to error with `Function does not return a value (it only ever returns None)  [func-returns-value]` on the return statement